flink-connector-elasticsearch8/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch8Writer.java

```diff
@@ -0,0 +1,134 @@
+
+    private final ElasticsearchAsyncClient esClient;
+
+    public Elasticsearch8Writer(
+        ElementConverter<InputT, Operation> elementConverter,
+        Sink.InitContext context,
+        int maxBatchSize,
+        int maxInFlightRequests,
+        int maxBufferedRequests,
+        long maxBatchSizeInBytes,
+        long maxTimeInBufferMS,
+        long maxRecordSizeInBytes,
+        String username,
+        String password,
+        HttpHost[] httpHosts,
+        Collection<BufferedRequestState<Operation>> state
+    ) {
+        super(
+            elementConverter,
+            context,
+            AsyncSinkWriterConfiguration.builder()
+                .setMaxBatchSize(maxBatchSize)
+                .setMaxBatchSizeInBytes(maxBatchSizeInBytes)
+                .setMaxInFlightRequests(maxInFlightRequests)
+                .setMaxBufferedRequests(maxBufferedRequests)
+                .setMaxTimeInBufferMS(maxTimeInBufferMS)
+                .setMaxRecordSizeInBytes(maxRecordSizeInBytes)
+                .build(),
+            state
+        );
+
+        this.esClient = new NetworkConfigFactory(httpHosts, username, password).create();
+    }
+
+    @Override
+    protected void submitRequestEntries(List<Operation> requestEntries, Consumer<List<Operation>> requestResult) {
+        LOG.debug("submitRequestEntries with {} items", requestEntries.size());
+
+        BulkListener<Operation> listener = new BulkListener<Operation>() {
+            @Override
+            public void beforeBulk(long executionId, BulkRequest request, List<Operation> contexts) {}
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, List<Operation> contexts, BulkResponse response) {
+                LOG.debug("Bulk request " + executionId + " completed");
+                ArrayList<Operation> retriableEntries = new ArrayList<>();
+
+                for (int i = 0; i < contexts.size(); i++) {
+                    BulkResponseItem item = response.items().get(i);
+                    Operation operation = contexts.get(i);
+
+                    if (item.error() != null) {
+                        LOG.error("Failed operation " + operation + " - " + item.error().reason());
+                    }
+
+                    if (item.error() != null && operation.isRetriable()) {
+                        retriableEntries.add(operation);
+                        operation.retry();
+                    }
+                }
+
+                requestResult.accept(retriableEntries);
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, List<Operation> contexts, Throwable failure) {
+                LOG.debug("Bulk request " + executionId + " failed", failure);
+                requestResult.accept(requestEntries);
+            }
+        };
+
+        try (BulkIngester<Operation> ingester = BulkIngester.of(b -> b
```

Comment:
I think this breaks the semantics of the `submitRequestEntries` operation, from the javadocs (there is also an example there):

>  NOTE: The client must support asynchronous requests and the method called to persist the records must asynchronously execute and return a future with the results of that request.

The `BulkIngester::close` will block and wait for all requests to be completed

Reply:
I believe you're right. Thank you for pointing it out! 🤝

I've made a couple of tests, and indeed the `BulkIngester::close` was blocking. I've refactored to use `BulkRequest`, which returns a `CompletableFuture<BulkResponse>` as the `submitRequestEntries` specifies and it worked like a charm.